src/index/bip352.h

```diff
@@ -0,0 +1,77 @@
+#ifndef BITCOIN_INDEX_BIP352_H
+#define BITCOIN_INDEX_BIP352_H
+
+#include <coins.h>
+#include <index/base.h>
+#include <pubkey.h>
+
+class COutPoint;
+class CBlockUndo;
+
+static constexpr bool DEFAULT_BIP352_INDEX{false};
+static constexpr bool DEFAULT_BIP352_CT_INDEX{false};
+
+/**
+ * This index is used to look up the tweaked sum of eligible public keys for a
+ * given transaction hash. See BIP352.
+ *
+ * Currently only silent payments v0 exists. Future versions may expand the
+ * existing index or create a (perhaps overlapping) new one.
+ */
+class BIP352Index final : public BaseIndex
+{
+public:
+    /**
+     * For each block hash we store and array of transactions. For each
+     * transaction we store:
+     * - the tweaked pubkey sum
+     * - the highest output amount (for dust filtering)
+     */
+    using tweak_index_entry = std::vector<std::pair<CPubKey, uint8_t>>;
+    static constexpr uint8_t dust_shift{4}; // Hexasat
+    /* Maximum dust threshold that we can filter */
+    static constexpr CAmount max_dust_threshold{CAmount(UINT8_MAX - 1) << dust_shift};
+
+    class DB;
+
+private:
+    const std::unique_ptr<DB> m_db;
+
+    /**  Whether this index has transaction cut-through enabled */
+    bool m_cut_through{false};
+
+    bool AllowPrune() const override { return false; }
+
+    /**
+     * Derive the silent payment tweaked public key for every block transaction.
+     *
+     *
+     * @param[in] txs all block transactions
+     * @param[in] block_undo block undo data
+     * @param[out] index_entry the tweaked public keys, only for transactions that have one
+     * @return false if something went wrong
+     */
+    bool GetSilentPaymentKeys(std::vector<CTransactionRef> txs, CBlockUndo& block_undo, tweak_index_entry& index_entry);
```

Comment:
The call signature of this method should be changed.. it;

1. needlessly rakes a copy of the read-only vector `txs`
2. needlessly requires a mutable `CBlockUndu&` which may lead the caller to think that his undo object is being modified by this method, when it is not.
3. This should be `const`

To accomplish (2) above, the called-function `BIP352::MakeSilentMayment` should be taking a `const CTransactionRef &`.

```suggestion
    bool GetSilentPaymentKeys(const std::vector<CTransactionRef>& txs, const CBlockUndo& block_undo, tweak_index_entry& index_entry) const;
```